### PR TITLE
Block vSphere on K8s 1.22

### LINKED ISSUE #7578: Block Kubernetes cluster 1.22 on vSphere
**User Story**
As reported in the CSI-driver documentation: 

> If `CSIMigrationvSphere` feature gate is enabled, users should not upgrade to Kubernetes v1.22. vSphere CSI Driver does not support Kubernetes v1.22 yet because it uses v1beta1 CRD APIs. Support for v1.22 will be added at a later release. Check the following document for supported Kubernetes releases for a given vSphere CSI Driver version.

Since #7280 introduced the external CCM for vSphere, and #7461 added the support for K8s v1.22, we need to avoid vSphere clusters using external CCM using K8S 1.22.
Below there are the needed steps to achieve this goal:
* Block upgrades for existing clusters to v1.22 if the `externalCloudProvider` feature is enabled
* Block CCM/CSI migration if the cluster version is already 1.22
* Block users to create 1.22 clusters on vSphere

**Acceptance criteria**
Creating clusters on vSphere with K8s v1.22 must not be possible.

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:
Since the necessity to block Kubernetes v1.22 on vSphere emerged under certain specific scenarios, this PR aims at implementing a generally available mechanism to block unwanted Kubernetes versions without the need to write custom code in the API server. The `extra-files` secret (where `versions.yaml` and `updates.yaml` reside) has been improved by introducing a new file: `provider-incompatibilities.yaml`, whose structure is the following:
```yaml
providerIncompatibilities:
- condition: always
  operation: CREATE
  provider: vsphere
  version: 1.22.*
- condition: externalCloudProvider
  operation: UPDATE
  provider: vsphere
  version: 1.22.*
```
Below are the parameters' details:
* condition: the condition that the cluster must match to trigger the incompatibility (e.g., the cluster using the external CCM)
* operation: the operation to be applied in (creation of a new cluster or update of an existing one)
* provider: the provider to which to apply the block
* version: the Kubernetes version to block (wildcards are allowed)

The conditions can be improved in the future, every field about the cluster or the datacenter is allowed (it's just a matter of a couple of lines of code).
Adding and removing constraints is very simple by using this mechanism: when in the future we'll allow using 1.22 clusters on vSphere, it will be enough to delete the above rules from the `defaults.go` [file](https://github.com/kubermatic/kubermatic/blob/06fa2a02e3fb09b70888a1795d24373f41199d6f/pkg/controller/operator/common/defaults.go#L265-L278), without dealing with ad-hoc conditions embedded in the core code.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7578 

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: charts/kubermatic/static/master/provider-incompatibilities.yaml
Comment (xrstf) [inline charts/kubermatic/static/master/provider-incompatibilities.yaml:1]: I appreciate the effort, but we do not update this chart anymore. It's soon going to be deleted entirely, so you can skip everything needed to generate this file.

Comment (xrstf) [reply charts/kubermatic/static/master/provider-incompatibilities.yaml:1]: NB: We _still_ required the seed-ctrlmgr to read data from files (instead of the `KubermaticConfiguration`), but work is being done to change that (i.e. we do not want to have a versions.yaml, updates.yaml etc. anymore, all we want is one single KubermaticConfiguration CRD and that's it).

So we might still need code to generate this file, but we do not update this Helm chart anymore. The KKP operator must create the file instead.

File: codegen/kubermatic_operator/main.go
Comment (xrstf) [inline codegen/kubermatic_operator/main.go:73]: i.e. this can go away

File: pkg/crd/operator/v1alpha1/configuration.go
Comment (xrstf) [inline pkg/crd/operator/v1alpha1/configuration.go]: Please document these fields and then regenerate the docs.

File: pkg/crd/operator/v1alpha1/configuration.go
Comment (xrstf) [inline pkg/crd/operator/v1alpha1/configuration.go]: > Wildcards are allowed, e.g. "1.22.*".

That is a statement that belongs on the "Version" field in the `Incompatibiliy` struct.

File: pkg/handler/v1/common/common.go
Comment (xrstf) [inline pkg/handler/v1/common/common.go]: Why did you decide to do it this way? Why not simply extend the existing `GetVersions` function? You were able to extend the `GetPossibleUpdates` already.

File: pkg/handler/v1/metering/metering.go
Comment (xrstf) [inline pkg/handler/v1/metering/metering.go]: :heart:

File: pkg/handler/v1/common/common.go
Comment (mlavacca) [reply pkg/handler/v1/common/common.go]: I did in this way because GetVersions is consumed by [this](https://github.com/kubermatic/kubermatic/blob/489c811eebf2419af4e83c8b1607fcd5a360daba/pkg/handler/routes_v1.go#L1668-L1689) endpoint, which has no way to fetch provider information, therefore must be replaced. I think it shouldn't be deleted before the UI has been migrated to use the new endpoint instead ([this](https://github.com/kubermatic/kubermatic/blob/1cd656ec7264744f3752b832273a43cf497ab211/pkg/handler/v2/routes_v2.go#L3554-L3579) one). Once The new endpoint will be correctly used by the UI, I think we can get rid of the old endpoint and change the getVersions function signature as you suggested.

File: pkg/handler/v1/metering/metering.go
Comment (mlavacca) [reply pkg/handler/v1/metering/metering.go]: Since when I discovered gimps, my life has changed :grin:

File: pkg/crd/operator/v1alpha1/configuration.go
Comment (mlavacca) [reply pkg/crd/operator/v1alpha1/configuration.go]: right, gonna change it

File: pkg/handler/v2/version/version.go
Comment (xrstf) [inline pkg/handler/v2/version/version.go]: If we already validated the request beforehand, what possible error conditions can happen here and are they really the user's fault (HTTP 4xx)?

File: pkg/handler/v2/version/version_test.go
Comment (xrstf) [inline pkg/handler/v2/version/version_test.go]: you probably meant to write "and _compatible_ with"

File: pkg/provider/types.go
Comment (xrstf) [inline pkg/provider/types.go]: A quick ctrl-f in Firefox did not show any spots where this is being used. Am I wrong? :)

File: pkg/resources/cloudcontroller/vsphere.go
Comment (xrstf) [inline pkg/resources/cloudcontroller/vsphere.go]: This is kind of a conflict for me. One the one hand, we make it configurable what is supported and what not, but this hardcodes it again to the KKP codebase. So which wins?

Can we rewrite this to make use of the configured incompatibilities? WDYT @xmudrii ? Technically this then also, in the future, affects other providers where we currently have hardcoded `switch` statements that technically would need to be moved to this new configuration file, right?

File: pkg/version/helpers.go
Comment (xrstf) [inline pkg/version/helpers.go]: What does this function create, besides a temporary variable that I, the caller, do not care in the slightest about? ^^ This could be shortened to just `IsConstrained`, no?

File: pkg/version/manager.go
Comment (xrstf) [inline pkg/version/manager.go]: Please document these and the other new constants. I am not even sure what exactly these two control ^^

**Review by xrstf — COMMENTED** [review]
Very nice work, yet I have a few questions and comments :)

File: pkg/provider/types.go
Comment (mlavacca) [reply pkg/provider/types.go]: No, you are right, probably it is garbage left from the previous idea that I had.

File: pkg/handler/v1/common/common.go
Comment (xrstf) [reply pkg/handler/v1/common/common.go]: I see. I am so scared that the two versions will live forever and we will never clean up... so scared.. can you calm me down by promising you will clean it up? ^^

Comment (mlavacca) [reply pkg/handler/v1/common/common.go]: Once the UI part will be implemented I'll clean it up, you have my word :grin:

File: pkg/resources/cloudcontroller/vsphere.go
Comment (mlavacca) [reply pkg/resources/cloudcontroller/vsphere.go]: I agree that having this hardcoded switch is in conflict with the goal of this PR. We can use the incompatibilities file also to check these kinds of conditions without too much effort if it worths it.

File: pkg/version/helpers.go
Comment (mlavacca) [reply pkg/version/helpers.go]: Definitely a smarter name, yes.

File: pkg/handler/v2/version/version.go
Comment (mlavacca) [reply pkg/handler/v2/version/version.go]: Actually, if an error occurs in that situation, there is some problem with the versions in the extra-files, therefore I replaced 400 with 500.

File: pkg/handler/common/cluster.go
Comment (xmudrii) [inline pkg/handler/common/cluster.go]: Does this apply to all providers or only to vSphere? I think allowing CCM/CSI migration for OpenStack 1.22 clusters should be fine.

NB: The CCM/CSI migration actually might not work properly since 1.21 because some feature gates got changed. I can take a look at this later.

File: pkg/handler/v1/common/common.go
Comment (xmudrii) [reply pkg/handler/v1/common/common.go]: Would it make sense to add a comment shortly describing this, maybe even linking to this thread? It might help if someone is reading this code later. :slightly_smiling_face:

File: pkg/resources/cloudcontroller/vsphere.go
Comment (xmudrii) [reply pkg/resources/cloudcontroller/vsphere.go]: Yes, I agree as well.

File: pkg/version/manager.go
Comment (xmudrii) [inline pkg/version/manager.go]: I find calling this `UPDATE` slightly confusing. As far as I get from the code, this is used to block cluster upgrades. Therefore, it might make sense to call it `UPGRADE` instead.

**Comment by xrstf** [comment]
I

/approve

and @xmudrii can give their /lgtm

Also, I

/retest

**Comment by mlavacca** [comment]
> I
> 
> /approve
> 
> and @xmudrii can give their /lgtm
> 
> Also, I
> 
> /retest

@xrstf  I'm implementing this part:
> Can we rewrite this to make use of the configured incompatibilities? WDYT @xmudrii ? Technically this then also, in the future, affects other providers where we currently have hardcoded switch statements that technically would need to be moved to this new configuration file, right?

should be ready in a while

File: charts/kubermatic/static/master/provider-incompatibilities.yaml
Comment (mlavacca) [reply charts/kubermatic/static/master/provider-incompatibilities.yaml:1]: @xrstf I had to recreate the file since it is used (along `with versions.yaml` and `updates.yaml`) by the `run-api.sh` script. If we don't statically provide this file to it, that test script is unusable.

File: pkg/resources/cloudcontroller/vsphere.go
Comment (mlavacca) [reply pkg/resources/cloudcontroller/vsphere.go]: @xmudrii @xrstf I used the incompatibilities mechanism to block also the migration, this ad-hoc part for vSphere has been deleted.

File: pkg/handler/common/cluster.go
Comment (mlavacca) [reply pkg/handler/common/cluster.go]: Solved by using the incompatibilities mechanism

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

**Comment by xrstf** [comment]
/hold

temporarily. I'm doing something on CI. Nothing to worry about.

**Comment by mlavacca** [comment]
@xmudrii LGTM lost after rebase (there was a conflict). Could you lgtm again please?

**Review by xmudrii — APPROVED** [review]

**Pull request merged**
